core/kernel/interrupt.c

```diff
@@ -23,27 +23,28 @@ void itr_init(struct itr_chip *chip)
 	itr_chip = chip;
 }
 
-static struct itr_handler *find_handler(size_t it)
-{
-	struct itr_handler *h;
-
-	SLIST_FOREACH(h, &handlers, link)
-		if (h->it == it)
-			return h;
-	return NULL;
-}
-
 void itr_handle(size_t it)
 {
-	struct itr_handler *h = find_handler(it);
-
-	if (!h) {
-		EMSG("Disabling unhandled interrupt %zu", it);
-		itr_chip->ops->disable(itr_chip, it);
-		return;
+	struct itr_handler *h = NULL;
+	enum itr_return ret = ITRR_NONE;
+	bool was_handled = false;
+
+	SLIST_FOREACH(h, &handlers, link) {
+		if (h->it != it)
+			continue;
+
+		ret = h->handler(h);
+
+		if (ret == ITRR_HANDLED) {
+			was_handled = true;
+		} else if (ret != ITRR_HANDLED && !(h->flags & ITRF_SHARED)) {
```

Comment:
You can drop `ret != ITRR_HANDLED` as we know that's the case here due to the `else`.

Reply:
yes, you are right, 
v4 is pushed 

Thanks,
Davidwang